applepay/utils.py

```diff
@@ -66,3 +100,316 @@ def signing_time_is_valid(signature, current_time, threshold):
         threshold=threshold)
     )
     return is_valid
+
+
+def valid_chain_of_trust(root_cert_der, intermediate_cert_der, leaf_cert_der):
+    """Validate the chain of trust for the provided der encoded root, intermediate,
+    and leaf certificates.
+
+    From: #ApplePaySpec
+       Part C: Ensure that there is a valid X.509 chain of trust from the
+       signature to the root CA. Specifically, ensure that the signature
+       was created using the private key corresponding to the leaf certificate,
+       that the leaf certificate is signed by the intermediate CA, and that the
+       intermediate CA is signed by the Apple Root CA - G3.
+
+    Args:
+        root_cert_der (String): der-encoded root cert
+        intermediate_cert_der (String): der-encoded intermediate cert
+        leaf_cert_der (String): der-encoded leaf cert
+    Returns:
+        Boolean: If there is a valid chain of trust for the provided certificates
+    """
+    root_cert = crypto.load_certificate(crypto.FILETYPE_ASN1, root_cert_der)
+
+    # Only add certs we TRUST!
+    store = crypto.X509Store()
+    store.add_cert(root_cert)
+
+    # validate the intermediate cert against root
+    intermediate_cert = crypto.load_certificate(crypto.FILETYPE_ASN1, intermediate_cert_der)
+    store_ctx = crypto.X509StoreContext(store, intermediate_cert)
+    # throws 'X509StoreContextError' when cert is invalid.
+    try:
+        store_ctx.verify_certificate()
+    except crypto.X509StoreContextError:
+        logger.warning("Intermediate cert not signed by the Apple Root CA - G3.")
+        return False
+    else:
+        store.add_cert(intermediate_cert)
+
+    # validate the leaf cert against intermediate.
+    leaf_cert = crypto.load_certificate(crypto.FILETYPE_ASN1, leaf_cert_der)
+    store_ctx = crypto.X509StoreContext(store, leaf_cert)
+    # throws 'X509StoreContextError' when cert is invalid.
+    try:
+        store_ctx.verify_certificate()
+    except crypto.X509StoreContextError:
+        logger.warning("Leaf cert not signed by the intermediate cert.")
+        return False
+    else:
+        store.add_cert(leaf_cert)
+
+    return True
+
+
+def get_leaf_and_intermediate_certs(candidate_certificates):
+    """Iterate over the provided candidate_certificates
+    to find the leaf and intermediate certificates based
+    on their OIDs.
+
+    Args:
+        candidate_certificate (asn1crypto.cms.CertificateSet): Iterable of
```

Comment:
It may be helpful to call out that `CertificateSet` subclasses `SetOf`, which is unordered.